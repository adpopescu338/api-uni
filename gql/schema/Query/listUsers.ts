import { gql } from 'graphql-tag';

export const listUsers = gql`
  type Query {
    """
    Only sysadmins can list users
    """
    listUsers(size: Int, page: Int): ListUsersResponse! @sysadmin
  }

  type ListUsersResponse {
    users: [User!]!
    total: Int!
    nextPage: Int
  }
`;
