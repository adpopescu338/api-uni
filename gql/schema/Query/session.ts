import { gql } from 'graphql-tag';

export const session = gql`
  type Query {
    """
    Get the current session details (no user details are returned apart from the user ID)
    This will return an error if the user is not logged in
    """
    session: Session! @authenticated
  }

  type Session {
    id: String!
    userId: String!
    createdAt: String!
    expiresAt: String!
  }
`;
