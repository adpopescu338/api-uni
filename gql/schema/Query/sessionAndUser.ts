import { gql } from 'graphql-tag';

export const sessionAndUser = gql`
  type Query {
    """
    Get the current session and user details
    """
    sessionAndUser: SessionAndUserResponse! @authenticated
  }

  type SessionAndUserResponse {
    session: Session!
    user: User!
  }
`;
