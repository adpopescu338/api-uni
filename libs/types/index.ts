export type { Context } from './Context';
export * from './generated';
export { GQLResolver } from './GqlResolver';
export { RedisSession } from './RedisSession';
