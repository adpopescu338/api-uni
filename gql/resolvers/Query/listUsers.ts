import { throwIfNotAdmin } from 'libs/throwIfNotAdmin';
import { Context } from 'libs/types';
import { ListUsersResponse, QueryListUsersArgs } from 'libs/types/generated';
import { UserIncludeRolesAndPermissions, userIncludeRolesAndPermissions } from 'prisma/selectors';

const DEFAULT_SIZE = 5;

export const listUsers = async (
  _: unknown,
  { page, size }: QueryListUsersArgs,
  ctx: Context,
): Promise<ListUsersResponse> => {
  throwIfNotAdmin(ctx.user);

  size ??= DEFAULT_SIZE;
  const [users, total]: [UserIncludeRolesAndPermissions[], number] = await Promise.all([
    ctx.prisma.user.findMany({
      skip: page ? (page - 1) * DEFAULT_SIZE : undefined,
      take: DEFAULT_SIZE,
      ...userIncludeRolesAndPermissions,
    }),
    countUsers(ctx),
  ]);

  const nextPage = users.length < size ? null : (page ?? 1) + 1;

  return {
    users,
    total,
    nextPage,
  } as ListUsersResponse;
};

const countUsers = async ({ prisma, cacheManager }: Context) => {
  const cachedCount = await cacheManager.get<number>(cacheManager.cacheKeys.usersCount);

  if (cachedCount === 0 || cachedCount) {
    return cachedCount;
  }

  const count = await prisma.user.count();

  // no need to await
  cacheManager.set(cacheManager.cacheKeys.usersCount, count);

  return count;
};
