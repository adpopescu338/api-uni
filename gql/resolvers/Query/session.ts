import { Context, RedisSession } from 'libs/types';
import { Session } from 'libs/types/generated';

export const session = async (
  _: unknown,
  __: unknown,
  { cacheManager, sessionKey }: Context,
): Promise<Session | null | undefined> => {
  if (!sessionKey) return null;

  const sessionData = await cacheManager.get<RedisSession>(sessionKey, true);

  return sessionData;
};
