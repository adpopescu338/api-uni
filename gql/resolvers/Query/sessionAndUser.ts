import { GraphQLError } from 'graphql';
import { Context, RedisSession } from 'libs/types';
import { SessionAndUserResponse } from 'libs/types/generated';

export const sessionAndUser = async (
  _: unknown,
  __: unknown,
  { cacheManager, prisma, sessionKey }: Context,
): Promise<SessionAndUserResponse | null> => {
  if (!sessionKey) return null;

  const sessionData = await cacheManager.get<RedisSession>(sessionKey, true);

  if (!sessionData) {
    throw new GraphQLError('Invalid session');
  }

  const user = await prisma.user.findUnique({
    where: {
      id: sessionData.user.id,
    },
    include: {
      roles: {
        include: {
          permissions: true,
        },
      },
    },
  });

  return {
    session: sessionData,
    user,
  } as SessionAndUserResponse;
};
