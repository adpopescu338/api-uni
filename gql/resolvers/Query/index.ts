import { listUsers } from './listUsers';
import { session } from './session';
import { sessionAndUser } from './sessionAndUser';

export const Query = { listUsers, session, sessionAndUser };
