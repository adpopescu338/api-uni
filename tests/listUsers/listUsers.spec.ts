import { describe, it, expect, jest } from '@jest/globals';
import { listUsers } from 'gql/resolvers/Query/listUsers';
import { GraphQLError } from 'graphql';
import { Context } from 'libs/types';
import { ctx, expectedResultWithCountFromCache, expectedResultWithCountFromPrisma } from './data';

describe('listUsers', () => {
  it('Should be defined', () => {
    expect(listUsers).toBeDefined();
  });

  it('Should return the expected result', async () => {
    const result = await listUsers(null, {}, ctx);

    expect((ctx.prisma.user.findMany as jest.Mock).mock.calls).toMatchSnapshot();
    expect(result).toEqual(expectedResultWithCountFromCache);
    expect(ctx.cacheManager.get).toHaveBeenCalledWith('usersCount');
    expect(ctx.prisma.user.count).not.toHaveBeenCalledTimes(1);
  });

  it('Should return the expected result with count from prisma', async () => {
    (ctx.cacheManager.get as jest.Mock).mockResolvedValueOnce(null as never);

    const result = await listUsers(null, {}, ctx);

    expect((ctx.prisma.user.findMany as jest.Mock).mock.calls).toMatchSnapshot();
    expect(result).toEqual(expectedResultWithCountFromPrisma);
    expect(ctx.cacheManager.get).toHaveBeenCalledWith('usersCount');
    expect(ctx.prisma.user.count).toHaveBeenCalled();
    expect(ctx.cacheManager.set).toHaveBeenCalledWith(
      'usersCount',
      expectedResultWithCountFromPrisma.total,
    );
  });
});
