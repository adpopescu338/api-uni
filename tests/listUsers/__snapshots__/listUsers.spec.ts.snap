// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`listUsers Should return the expected result 1`] = `
[
  [
    {
      "include": {
        "roles": {
          "include": {
            "permissions": true,
          },
        },
      },
      "skip": undefined,
      "take": 5,
    },
  ],
]
`;

exports[`listUsers Should return the expected result with count from prisma 1`] = `
[
  [
    {
      "include": {
        "roles": {
          "include": {
            "permissions": true,
          },
        },
      },
      "skip": undefined,
      "take": 5,
    },
  ],
  [
    {
      "include": {
        "roles": {
          "include": {
            "permissions": true,
          },
        },
      },
      "skip": undefined,
      "take": 5,
    },
  ],
]
`;
